core/domain/classifier_services.py

```diff
@@ -127,6 +129,117 @@ def classify_string_classifier_rule(state, normalized_answer):
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+
+        # Validate the job.
+        dummy_classifier_training_job = classifier_domain.ClassifierTrainingJob(
+            'job_id_dummy', algorithm_id, interaction_id, exp_id, exp_version,
+            state_name, feconf.TRAINING_JOB_STATUS_NEW, training_data)
+        dummy_classifier_training_job.validate()
+
+        job_dicts_list.append({
+            'algorithm_id': algorithm_id,
+            'interaction_id': interaction_id,
+            'exp_id': exp_id,
+            'exp_version': exp_version,
+            'state_name': state_name,
+            'training_data': training_data,
+            'status': feconf.TRAINING_JOB_STATUS_NEW
+        })
+
+    # Create all the classifier training jobs.
+    job_ids = classifier_models.ClassifierTrainingJobModel.create_multi(
+        job_dicts_list)
+
+    # Create mapping for each job.
+    job_exploration_mappings = []
+    for job_id_index, job_id in enumerate(job_ids):
+        job_exploration_mapping = (
+            classifier_domain.TrainingJobExplorationMapping(
+                job_dicts_list[job_id_index]['exp_id'],
+                job_dicts_list[job_id_index]['exp_version'],
+                job_dicts_list[job_id_index]['state_name'],
+                job_id))
+        job_exploration_mapping.validate()
+        job_exploration_mappings.append(job_exploration_mapping)
+
+    classifier_models.TrainingJobExplorationMappingModel.create_multi(
+        job_exploration_mappings)
+
+
+def handle_non_retrainable_states(exploration, state_names,
+                                    new_to_old_state_names):
+    """Creates new TrainingJobExplorationMappingModel instances for all the
+    state names passed into the function. The mapping is created from the
+    new state to the ClassifierTrainingJob of the old state. If there's been
+    a change in the state name, we retrieve the old state name and create the
+    mapping accordingly.
+    This method is called only from exp_services._save_exploration() method and
+    is never called from exp_services._create_exploration().
+
+    Args:
+        exploration: Exploration. The Exploration domain object.
+        state_names: list(str). List of state names.
+        new_to_old_state_names: dict. Dict mapping new state names to their
+            corresponding state names in previous version.
+
+    Raises:
+        Exception. This method should not be called by exploration with version
+            number 1.
+    """
+    exp_id = exploration.id
+    current_exp_version = exploration.version
+    old_exp_version = current_exp_version - 1
+    if old_exp_version <= 0:
+        raise Exception(
+            'This method should not be called by exploration with version '
+            'number %s' % (current_exp_version))
+
+    state_names_to_retrieve = []
+    for current_state_name in state_names:
```

Comment:
@anmolshkl Are you sure we should do that? This makes the mapping variable's name really long as well. Do you think adding a doc string explaining this would suffice?